Integer Scaling support implementation to improve the viewing quality of Pixel Art games

slattann opened the pull request:
**Background:**
Blurry outputs during up-scaling the video buffer, is a generic problem
of graphics industry. One of the major reason behind this blurriness is
the interpolation of pixel values used by most of the up-scaling hardware.

Nearest-neighbor is a scaling filter type, which works by filling in the
missing color values in the up-scaled image with that of the coordinate-mapped
nearest source pixel value.

Nearest-neighbor can produce (almost) non-blurry scaling outputs when
the scaling ratio is complete integer. For example:

input buffer resolution: 1280x720(HD)
output buffer resolution: 3840x2160(UHD/4K)
scaling ratio (h) = 3840/1280 = 3
scaling ratio (v) = 2160/720 = 3
In such scenarios, we should try to pick Nearest-neighbor as scaling
method when possible.

**Some references:**

- https://tanalin.com/en/articles/integer-scaling/
- 
- https://www.reddit.com/r/nvidia/comments/9e27wk/its_2018_nvidia_wheres_integer_scaling/
- 
- http://www.zeus-software.com/forum/viewtopic.php?t=2065
- 
- https://forums.geforce.com/default/topic/844905/geforce-drivers/-feature-request-nonblurry-upscaling-at-integer-ratios/20/
- 
- https://forums.tomshardware.com/threads/does-anybody-have-a-good-solution-to-scale-1080p-onto-a-4k-monitor.3217086/
- 

**This patch series:**
To address this feature kernel driver exposes new CRTC/Plane property called
"SCALING_FILTER" and gives an option to userspace to set it based on the use
cases.
Kernel RFC can be found here: https://patchwork.freedesktop.org/series/73884/

This patchset add support for utilizing CRTC/Plane scaling filter property based on
client's usecase.

Currently used some hack to input resolution hard coded eg: 640x480 instead of setting based on the CRTC mode.

**Here are the patches details:**
Patch 1/3 --> Disable GL-Scaling to make sure that scalling is done in DE via DRM path
Patch 2/3 --> add scaling_filter property support at plane level
Patch 3/3 --> Hacked by hard coding the source resolution info

**Kindly help us do it properly w/o hacking the input resolutions.**
-------------------------------------------------------------------------------------
Here are some sample reference data for having Integer scaling with NN filter:

OS Res X | OS Res Y | Panel X | Panel Y | Aspected Scaled dest X | Aspected Scaled dest Y | Scaled ratio X | Scaled ratio Y | IS dest X | IS dest Y
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
640 | 480 | 3840 | 2160 | 2880 | 2160 | 6 | 4.5 | 2560 | 1920
800 | 600 | 3840 | 2160 | 2880 | 2160 | 3.6 | 3.6 | 2400 | 1800
1024 | 768 | 3840 | 2160 | 2880 | 2160 | 2.8125 | 2.8125 | 2048 | 1536
1280 | 720 | 3840 | 2160 | 3840 | 2160 | 3 | 3 | 3840 | 2160
1280 | 1024 | 3840 | 2160 | 2700 | 2160 | 2.1094 | 2.1094 | 2560 | 2048
1920 | 1080 | 3840 | 2160 | 3840 | 2160 | 2 | 2 | 3840 | 2160
1920 | 1200 | 3840 | 2160 | 3456 | 2160 | 1.8 | 1.8 | 1920 | 1200

Awaiting your reviews and what do you think on having Integer Scaling in Kodi. Thanks
-Sameer Lattannavar

slattann committed:
[GL-Scaling]: disabling GL-Scaling

As part of
 https://software.intel.com/en-us/articles/integer-scaling-support-on-intel-graphics
 work, enabling integer scaling in middelware.

To make DRM scaling work, disabling GL-scaling by setting
needsScaling = false and relevant changes.

slattann committed:
[CDRMAtomic]: add scaling_filter property support at plane level.

As part of https://software.intel.com/en-us/articles/integer-scaling-support-on-intel-graphics
this work, enabling integer scaling for better Pixel Art games at plane level.

This commit has 4 parts to it:
1) adding SCALING_FILTER property support at plane level only when
user toggles Setting UseScalingFilter at Settings->System->Display
2) calculating scaling factor based on source & destination
resolutions with maintaing the aspect ratio.
3) reading DRM property for right Scaling Filter eg: "Nearest Neighbor"
4) retaining default behavior when user does not toggle the setting

Ref:Kernel changes are floated at https://patchwork.freedesktop.org/series/73883/#rev6

slattann committed:
[HACK]GBMUtils: hacked by hard coding the source resolution info

As part of
 https://software.intel.com/en-us/articles/integer-scaling-support-on-intel-graphics
 work, enabling integer scaling in middelware.

This commit hacks smaller resolution of surface before calling,
CreateSurface().

emersion commented on xbmc/windowing/gbm/DRMAtomic.cpp:
```diff
@@ -58,14 +66,30 @@ void CDRMAtomic::DrmAtomicCommit(int fb_id, int flags, bool rendered, bool video
+	AddProperty(m_gui_plane, "SCALING_FILTER", property_id ? 1 : 0);
+	property_id = this->GetPropertyId(m_video_plane, "SCALING_FILTER");
+	CLog::Log(LOGINFO, "SAMEER --> [%s:%d] Property_ID:%ld - SCALING_FILTER\n", __func__, __LINE__,property_id );
+	AddProperty(m_video_plane, "SCALING_FILTER", property_id ? 1 : 0);
```
This hardcodes the `SCALING_FILTER` to 1 if the property is advertised.

- What does this value mean?
- If the property doesn't exist, shouldn't we **not** set it?
- Is an hardcoded value enough to demonstrate that the new `SCALING_FILTER` property works?

ppaalanen replied:
Furthermore, using hardcoded values 0 and 1 is incorrect.

Userspace must query the kernel what values the strings `"Default"` and `"Nearest Neighbor"` map to, and then use those values. Userspace *must not* assume the values can be hardcoded.

slattann replied:
Thanks for your comments, I will rework on those. I had plan to add enums to be more meaning.
I will consider your input and work on that.

slattann replied:
@emersion/ppaalanen,

From kernel I can query only current filter value (default) and it does not give me complete supported list of filters. Is it fine to define our own enum inline with kernel enum?
Does it make sense ?

enum drm_scaling_filter {
   DRM_SCALING_FILTER_DEFAULT,
    DRM_SCALING_FILTER_NEAREST_NEIGHBOR,
};
.
.
.
if(SupportsProperty(m_video_plane, "SCALING_FILTER")){
	AddProperty(m_video_plane, "SCALING_FILTER", DRM_SCALING_FILTER_NEAREST_NEIGHBOR);
}


Refer Kernel patch series @: https://patchwork.freedesktop.org/series/73883/#rev6


-Sameer Lattannavar

ppaalanen replied:
No, under the current kernel UAPI rules, it is not fine.

You can query the kernel for the whole list though: `drmModeObjectGetProperties()` gives you everything you need to know.

lrusak replied:
I imagine something like this would be what you are looking for? https://github.com/lrusak/xbmc/commit/9eac28915d8a86b9c42bd9fac2c43e555e222931

ppaalanen replied:
Yes, that has the right idea.

slattann replied:
Thanks @ppaalanen & @lrusak for your inputs. Please suggest me if these changes are fine?
 ----------------------------------------------------------------------------------------------------
if(CServiceBroker::GetWinSystem()->UseScalingFilter()){
     if(SupportsProperty(m_gui_plane, "SCALING_FILTER")){
		uint32_t prop_id = this->GetPropertyId(m_gui_plane, "SCALING_FILTER");
		drmModePropertyPtr m_gui_prop = drmModeGetProperty(m_fd, prop_id);
		for(int i=0; i<m_gui_prop->count_enums; i++){
				if(!strcmp(m_gui_prop->enums[i].name, "Nearest Neighbor"))
						AddProperty(m_gui_plane, "SCALING_FILTER", m_gui_prop->enums[i].value);
		}
		}
        if(SupportsProperty(m_video_plane, "SCALING_FILTER")){
			uint32_t prop_id = this->GetPropertyId(m_video_plane, "SCALING_FILTER");
			drmModePropertyPtr m_video_prop = drmModeGetProperty(m_fd,prop_id);
			for(int i=0; i<m_video_prop->count_enums; i++){
					if(!strcmp(m_video_prop->enums[i].name, "Nearest Neighbor"))
							AddProperty(m_video_plane, "SCALING_FILTER", m_video_prop->enums[i].value);
			}
}
-Sameer Lattannavar

lrusak commented:
At a minimum (without doing a full review) the option to use the new integer scaling should be made selectable via a setting in the system settings. This way it doesn't have to be forced upon users.

Also as already said by others, you should query the plane to make sure it supports that property before just blindly trying to set it 👍 You can do that via the `SupportsProperty` method https://github.com/xbmc/xbmc/blob/master/xbmc/windowing/gbm/DRMUtils.h#L120

I hope that gives you a bit to go on. If you need help creating the setting please just ask 😄

slattann committed:
[Settings]: adding UseScalingFilter option into setting menu as a toggle

Adding ScalingFilter toggle into Settings->system->Display as
UseScalingFilter.

This patch has 2 parts:
1) Setting toggle of UseScalingFilter into gbm.xml
2) Implementation of using the setting while making decision
to set ScalingFilter into DRM

slattann committed:
incorrect 'object' rectified.

lrusak commented on xbmc/settings/Settings.cpp:
```diff
@@ -418,6 +418,7 @@ const std::string CSettings::SETTING_SOURCE_VIDEOS = "source.videos";
 const std::string CSettings::SETTING_SOURCE_MUSIC = "source.music";
 const std::string CSettings::SETTING_SOURCE_PICTURES = "source.pictures";
 
+const std::string CSettings::SETTING_VIDEOSCREEN_SCALING_FILTER="videoscreen.scalingfilter";
```
This can exist in the file it's used (`WinSystemGbm.cpp`) we do this to avoid polluting Settings.cpp with platform specifics

lrusak commented on xbmc/settings/Settings.h:
```diff
@@ -386,6 +386,7 @@ class CSettings : public CSettingsBase, public CSettingCreator, public CSettingC
   static const int VIDEOLIBRARY_PLOTS_SHOW_UNWATCHED_TVSHOWEPISODES = 1;
   static const int VIDEOLIBRARY_THUMB_SHOW_UNWATCHED_EPISODE = 2;
 
+  static const std::string SETTING_VIDEOSCREEN_SCALING_FILTER;
```
unneeded, see above comment

lrusak commented on xbmc/windowing/gbm/DRMAtomic.cpp:
```diff
@@ -64,8 +84,15 @@ void CDRMAtomic::DrmAtomicCommit(int fb_id, int flags, bool rendered, bool video
       AddProperty(m_gui_plane, "CRTC_Y", 0);
-      AddProperty(m_gui_plane, "CRTC_W", m_mode->hdisplay);
-      AddProperty(m_gui_plane, "CRTC_H", m_mode->vdisplay);
+      if (CServiceBroker::GetWinSystem()->UseScalingFilter()){
```
As this is specific to GBM windowing I wouldn't call into winsystem for this and just make it specific to CDRMAtomic

lrusak commented on xbmc/windowing/gbm/DRMAtomic.cpp:
```diff
@@ -22,11 +22,31 @@
+}
+bool CDRMAtomic::SetScalingFilter(struct drm_object *object, const char *name) {
+  uint32_t scale_factor;
+  for (uint32_t i = 0; i < object->props->count_props; i++)
```
Please use the `SupportsProperty` and `SupportsPropertyAndValue` helper functions here

lrusak reviewed (changes requested):
Other than some of the small changes I commented on you really need to use clang-format on the code changes. The formatting needs a lot of work and clang-format will take care of that automatically. I recommend setting up the git-clang-format hook so it runs anytime you make a new commit.

I personally use this one https://github.com/barisione/clang-format-hooks

Other than that I don't really know what implications the code changes have on other platforms that cannot use this code. This will have to be explored to make sure we don't break other platfoms.

lrusak commented:
One last thing, would it make sense to only use this when retroplayer is being used? I don't know if there is a benefit to using all the time? As Kodi has refresh rate switching for video playback it's simple to change the tv resolution to the media resolution to allow the tv to do the upscaling. This feature doesn't exist in retroplayer though.

contactshashanksharma commented:
I agree with @lrusak . Integer scaling can cause reduced clarity in some scenarios (like while displaying a font engine output from a browser), so it should be limited to the gaming scenarios, which is the actual aim of the feature too. It would make more sense to get the context of operation first and enable the property only when we are in the gaming context.

eigendude commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -21955,3 +21955,15 @@ msgstr ""
+#. Description of setting with label #39119 "Enable Integer Scaling Filter (NN)"
+#: system/settings/settings.xml
+msgctxt "#39120"
+msgid "Integer scaling (IS) is a nearest-neighbor upscaling technique that simply scales up the existing pixels by an integer (i.e., whole number) multiplier.Nearest-neighbor (NN) interpolation works by filling in the missing color values in the upscaled image with that of the coordinate-mapped nearest source pixel value"
```
I would propose a slight simplification (and space between sentences) of the help text.

Before:

> Integer scaling (IS) is a nearest-neighbor upscaling technique that simply scales up the existing pixels by an integer (i.e., whole number) multiplier.Nearest-neighbor (NN) interpolation works by filling in the missing color values in the upscaled image with that of the coordinate-mapped nearest source pixel value"

After:

> Integer scaling (IS) is an upscaling technique that scales up the existing pixels by a whole number. Nearest-neighbor (NN) interpolation works by filling in the missing color values with that of the nearest source pixel value.

garbear commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -21955,3 +21955,15 @@ msgstr ""
+#. Description of setting with label #39119 "Enable Integer Scaling Filter (NN)"
+#: system/settings/settings.xml
+msgctxt "#39120"
+msgid "Integer scaling (IS) is a nearest-neighbor upscaling technique that simply scales up the existing pixels by an integer (i.e., whole number) multiplier.Nearest-neighbor (NN) interpolation works by filling in the missing color values in the upscaled image with that of the coordinate-mapped nearest source pixel value"
```
I would propose a slight simplification (and space between sentences) of the help text.

Before:

> Integer scaling (IS) is a nearest-neighbor upscaling technique that simply scales up the existing pixels by an integer (i.e., whole number) multiplier.Nearest-neighbor (NN) interpolation works by filling in the missing color values in the upscaled image with that of the coordinate-mapped nearest source pixel value"

After:

> Integer scaling (IS) is an upscaling technique that scales up the existing pixels by a whole number. Nearest-neighbor (NN) interpolation works by filling in the missing color values with that of the nearest source pixel value.

contactshashanksharma replied:
I think the emphasis here is that 'IS' is a special case of the nearest-neighbor (NN) scaling technique. The nearest-neighbor technique works by generating the missing pixels information from neighboring pixels. Now, in the case of IS, as the target resolution is Integer times that of the source resolution, the whole pixel can be repeated, giving better clarity in the output.

garbear replied:
You're right, better would be:

> Integer scaling (IS) is a nearest-neighbor upscaling technique

While I oversimplified, I think when it comes to reading text on a TV, there's more value in less.

contactshashanksharma replied:
Agree

slattann commented:
I have posted a pull request at: https://github.com/xbmc/xbmc/pull/18194
Kindly review and give your feedback. Thanks
-Sameer Lattannavar

eigendude commented:
I'm a bit busy with work this week but can most likely find time this weekend to do some runtime testing. I'll send you any improvements I see.

garbear commented:
I'm a bit busy with work this week but can most likely find time this weekend to do some runtime testing. I'll send you any improvements I see.
